pkg/scheduler/queue/leafqueue.go

```diff
@@ -0,0 +1,137 @@
+package queue
+
+import (
+	"fmt"
+	"strings"
+)
+
+type QueuePath []string //nolint:revive
+
+// LeafQueue is an hierarchical queue implementation where each sub-queue
+// has the same guarantees to be chosen from.
+// Each queue has also a local queue, which gets chosen from first. Only if the
+// local queue is empty, items from the sub-queues are dequeued.
```

Comment:
Finding which item to dequeue looks kind like DFS to me. is my understanding right?

Reply:
Not sure if you can compare it with depth first search. 

My understanding of DFS would be that it would only pick the first item from the local queue, then pick from the subqueus. Only when all subqueues are dequeued, it dequeues from the local queue again.

However, here it dequeues all items from the local queue first, and only then proceeds to the subqueues.